### PR TITLE
Add seats to event types

### LINKED ISSUE #2064: RFC: EventTypes with slots/seats
## RFC: Event Types with slots/seats first steps
The following are all the description and ideas gathered for this new feature.

### Example User cases
- Private tutor can be booked "5 times per day" he has a 9 to 5 availability. He has 8 one-hour slots. But if he has been booked 5 times already, then he's done for the day and should appear unavailable even if he has slots left.
- Classroom has a 10 students per class limit. The time slot can be booked 10 times, one per student. After that the time slot can't be booked anymore but you can book other time slots.
- A conference-like event is taking place, it only has one time slot and can be booked 100 times. Each individual can book up to 5 seats. Once the 100 seats are booked the time slot is closed.

### Action Items

| Item             | Description                                                                       |
| ---------------- | --------------------------------------------------------------------------------- |
| seatsPerTimeSlot | max. quantity an event is allowed to be booked per individuals for each time slot |
| seatsPerAttendee | max. quantity an attendee is allowed to book for a time slot in a certain event   |
| bookingLimit     | max. quantity an event is available to be booked for a day/month/week             |
| bookingLimitType | string enum for type on bookingLimit (day/month/week)                             |

### Describe the solution you'd like

We have identified the next key features to develop:

- User can edit their **seatsPerTimeSlot** per event on the event type advanced settings section **(Required)**
- User can set an amount of _**seatsPerAttendee**_ for each booking. **(Required)**
- 5 or more seats availability is a PRO feature **(Required)**
- Attendee can book on an event based on its seat availability. **(Required)**
- Attendee can book up to _**seatsPerAttendee**_ and pay accordingly to number of seats selected **(Required)**
- Seat availability must be visible (Optional)
- Queue system, so if an attendee cancels users are moved within the queue (Optional)
- ~~User can set a window time availability for booking a event (Optional)~~ 

There was an attempt to this feature on PR #271, but if any challenge or something that can block development of this feature comes to your mind, let me know.

### Challenges
- How we should handle stripe and multiple seats selected? Payment, Invoice, Cancellation.
- What should happen when certain booking it's at its capacity and the User owner change slot availability to a less than previous max seats value?
- What approach to take for notifying User, when his event has reached capacity or all events available are booked?
- How synced calendars should display every booking made for the time-slot? Creation, updating, guest visibility for every calendar integration

### Opportunities 
- Send to the attendee a branded ticket number _# 001, # 002, ...# 006_ with a QR just like NextJs does on their events.
- Queue system in case there are no seats left.

@emrysal @alishaz-polymath @zomars

### PR DESCRIPTION
**Pull request by joeauyeung**
## What does this PR do?

This PR is a rework of #2392 since the commits I brought in made the branch messy. 

The previously reviewed PR's associated with seats are:
- #2370
- #2311
- #2392

Fixes #2064

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How should this be tested?

- [ ] Test A
- [ ] Test B

## Checklist

- I haven't checked if my PR needs changes to the documentation
- I haven't checked if my changes generate no new warnings
- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by joeauyeung** [comment]
@alannnc and @PeerRich tagged you two for review but the previous PRs that made this one up have already been reviewed.

**Comment by PeerRich** [comment]
> @alannnc and @PeerRich tagged you two for review but the previous PRs that made this one up have already been reviewed.

built seems failing on vercel

**Comment by joeauyeung** [comment]
@PeerRich builds now working on Vercel

File: packages/prisma/schema.prisma [hunk 1]
```diff
@@ -29,6 +29,12 @@ enum PeriodType {
   RANGE      @map("range")
 }
 
+enum BookingLimitType {
+  MONTH  @map("month")
+  DAY    @map("day")
+  WEEK   @map("week")
+}
+
```
Comment (zomars) [inline packages/prisma/schema.prisma:37]: This value isn't being used anywhere?

**Review by zomars — DISMISSED** [review]
Code looks good, there's still a couple of things missing.

Since this is a new type of event, E2E are needed to make sure this keeps working.

Also we need to see how this works with:
- Paid bookings
- Recurring events (recently added)
- Hashed Links (recently added)

File: packages/prisma/schema.prisma (hunk 1 above)
Comment (joeauyeung) [reply packages/prisma/schema.prisma]: Removed

**Comment by zomars** [comment]
image

We need to make some mobile adjustments

File: apps/web/lib/hooks/useSlots.ts [hunk 2]
```diff
@@ -56,6 +60,11 @@ export const getFilteredTimes = (props: getFilteredTimesProps) => {
       const slotStartTime = times[i];
       const slotEndTime = times[i].add(eventLength, "minutes");
       const slotStartTimeWithBeforeBuffer = times[i].subtract(beforeBufferTime, "minutes");
+      // If the event has seats then see if there is already a booking (want to show full bookings as well)
+      if (currentSeats?.some((booking) => booking.startTime === slotStartTime.toISOString())) {
+        console.log("This triggered");
```
Comment (zomars) [inline apps/web/lib/hooks/useSlots.ts:65]: We may want to log something more useful or remove it altogether

File: apps/web/lib/hooks/useSlots.ts [hunk 3]
```diff
@@ -124,26 +133,36 @@ export const useSlots = (props: UseSlotsProps) => {
 
     const handleAvailableSlots = async (res: Response) => {
       const responseBody: AvailabilityUserResponse = await res.json();
+      const { workingHours, busy, currentSeats } = responseBody;
       const times = getSlots({
         frequency: slotInterval || eventLength,
         inviteeDate: date,
-        workingHours: responseBody.workingHours,
+        workingHours: workingHours,
```
Comment (zomars) [inline apps/web/lib/hooks/useSlots.ts:140]:
```suggestion
        workingHours,
```

File: apps/web/lib/hooks/useSlots.ts [hunk 4]
```diff
@@ -124,26 +133,36 @@ export const useSlots = (props: UseSlotsProps) => {
...
         minimumBookingNotice,
         eventLength,
       });
       const filterTimeProps = {
         times,
-        busy: responseBody.busy,
+        busy: busy,
```
Comment (zomars) [inline apps/web/lib/hooks/useSlots.ts:146]:
```suggestion
        busy,
```

**Review by zomars — COMMENTED** [review]
We're almost there. Tested the preview and is working flawlessly. Nice job! Left some mobile and code nits.

**Review by PeerRich — APPROVED** [review]
lets goo

**Review by alannnc — APPROVED** [review]
It's working, and also edit booking also works with all guests.

**Pull request merged**